### PR TITLE
[spirv] Process clamp intrinsic function.

### PR DESCRIPTION
**Pull request by ehsannas**
Uses SPIR-V Extended Instructions for GLSL.

### CODE REVIEW

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 1]
```diff
@@ -2543,6 +2546,69 @@ uint32_t SPIRVEmitter::processIntrinsicCallExpr(const CallExpr *callExpr) {
...
+
+  // Get the function parameters. Expect 3 parameters.
+  assert(callExpr->getNumArgs() == 3u);
+  const Expr *argX = callExpr->getArg(0);
+  const Expr *argMin = callExpr->getArg(1);
+  const Expr *argMax = callExpr->getArg(2);
+  const QualType argXType = argX->getType();
+  const QualType argMinType = argMin->getType();
+  const QualType argMaxType = argMax->getType();
+  if (isFloatOrVecMatOfFloatType(argXType) ||
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: I think we only need to check the return type of the function call? The front end should've selected the correct overload already.

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 2]
```diff
@@ -2543,6 +2546,69 @@ uint32_t SPIRVEmitter::processIntrinsicCallExpr(const CallExpr *callExpr) {
   return processIntrinsicUsingGLSLInst(callExpr, glslOpcode, actOnEachVecInMat);
 }
 
+uint32_t SPIRVEmitter::processIntrinsicClamp(const CallExpr *callExpr) {
+  // According the HLSL reference: clamp(X, Min, Max) takes 3 arguments. Each
+  // one may be int, uint, or float.
+  // If *any* of the 3 arguments is a float, clamp will have the following
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done.

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp (hunk 1 above)
Comment (ehsannas) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Yep, that's a good idea.

File: tools/clang/lib/SPIRV/SPIRVEmitter.cpp [hunk 3]
```diff
@@ -2543,6 +2546,69 @@ uint32_t SPIRVEmitter::processIntrinsicCallExpr(const CallExpr *callExpr) {
...
+  if (isFloatOrVecMatOfFloatType(argXType) ||
+      isFloatOrVecMatOfFloatType(argMinType) ||
+      isFloatOrVecMatOfFloatType(argMaxType)) {
+    glslOpcode = GLSLstd450::GLSLstd450FClamp;
+  } else if (isSintOrVecMatOfSintType(argXType) ||
+             isSintOrVecMatOfSintType(argMinType) ||
+             isSintOrVecMatOfSintType(argMaxType)) {
+    glslOpcode = GLSLstd450::GLSLstd450SClamp;
+  } else {
+    glslOpcode = GLSLstd450::GLSLstd450UClamp;
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/SPIRVEmitter.cpp]: Done.

**Pull request merged**